shell/platform/embedder/embedder_surface_gl_impeller.cc

```diff
@@ -0,0 +1,195 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/embedder/embedder_surface_gl_impeller.h"
+
+#include <utility>
+
+#include "flutter/shell/common/shell_io_manager.h"
+#include "impeller/entity/gles/entity_shaders_gles.h"
+#include "impeller/renderer/backend/gles/context_gles.h"
+#include "impeller/renderer/backend/gles/proc_table_gles.h"
+#include "impeller/scene/shaders/gles/scene_shaders_gles.h"
+
+namespace flutter {
+
+class ReactorWorker final : public impeller::ReactorGLES::Worker {
+ public:
+  ReactorWorker() = default;
+
+  // |ReactorGLES::Worker|
+  bool CanReactorReactOnCurrentThreadNow(
+      const impeller::ReactorGLES& reactor) const override {
+    impeller::ReaderLock lock(mutex_);
+    auto found = reactions_allowed_.find(std::this_thread::get_id());
+    if (found == reactions_allowed_.end()) {
+      return false;
+    }
+    return found->second;
+  }
+
+  void SetReactionsAllowedOnCurrentThread(bool allowed) {
+    impeller::WriterLock lock(mutex_);
+    reactions_allowed_[std::this_thread::get_id()] = allowed;
+  }
+
+ private:
+  mutable impeller::RWMutex mutex_;
+  std::map<std::thread::id, bool> reactions_allowed_ IPLR_GUARDED_BY(mutex_);
+
+  FML_DISALLOW_COPY_AND_ASSIGN(ReactorWorker);
+};
+
+EmbedderSurfaceGLImpeller::EmbedderSurfaceGLImpeller(
+    EmbedderSurfaceGL::GLDispatchTable gl_dispatch_table,
+    bool fbo_reset_after_present,
+    std::shared_ptr<EmbedderExternalViewEmbedder> external_view_embedder)
+    : gl_dispatch_table_(std::move(gl_dispatch_table)),
+      fbo_reset_after_present_(fbo_reset_after_present),
+      external_view_embedder_(std::move(external_view_embedder)),
+      worker_(std::make_shared<ReactorWorker>()) {
+  worker_->SetReactionsAllowedOnCurrentThread(true);
```

Comment:
nit: should this be further down after the early returns?

Reply:
Done